README.md
==============================================================================
ReadMe file for the Experience-Based Language Acquisition (EBLA) system.
==============================================================================

$Id$


==============================================================================
LICENSE
==============================================================================

Copyright (c) 2002-2023, Brian E. Pangburn
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.  Redistributions in binary
form must reproduce the above copyright notice, this list of conditions and
the following disclaimer in the documentation and/or other materials
provided with the distribution.  The names of its contributors may not be
used to endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


==============================================================================
DESCRIPTION
==============================================================================

EBLA is an open computational framework for visual perception and
grounded language acquisition.  EBLA can watch a series of short
videos and acquire a simple language of nouns and verbs corresponding
to the objects and object-object relations in those videos.  Upon
acquiring this protolanguage, EBLA can perform basic scene analysis
to generate descriptions of novel videos.

The general architecture of EBLA is comprised of three stages: vision
processing, entity extraction, and lexical resolution.  In the vision
processing stage, EBLA processes the individual frames in short videos,
using a variation of the mean shift analysis image segmentation algorithm
to identify and store information about significant objects.  In the entity
extraction stage, EBLA abstracts information about the significant objects
in each video and the relationships among those objects into internal
representations called entities.  Finally, in the lexical acquisition
stage, EBLA extracts the individual lexemes (words) from simple
descriptions of each video and attempts to generate entity-lexeme mappings
using an inference technique called cross-situational learning.  EBLA is
not primed with a base lexicon, so it faces the task of bootstrapping its
lexicon from scratch.

While there have been several systems capable of learning object or event
labels for videos, EBLA is the first known system to acquire both nouns and
verbs using a grounded computer vision system.

EBLA was developed as part of Brian E. Pangburn's dissertation research
in the Department of Computer Science at Louisiana State University.

A brief workshop paper on EBLA is available from:
https://aclanthology.org/W03-0607.pdf

The full dissertation is available from:
https://digitalcommons.lsu.edu/gradschool_dissertations/2162/

Much of the vision system is based on a partial Java port of the EDISON
image processing system developed by Chris M. Christoudias and Bogdan
Georgescu at the Robust Image Understanding Laboratory at Rutgers University
(https://web.archive.org/web/20100814112001/http://coewww.rutgers.edu/riul/research/code/EDISON/index.html).
See the "seg_readme.txt" file for more information.

The graphical user interface (GUI) for EBLA was developed in part by
Prasanth R. Pasala.


==============================================================================
CODE EXECUTION, DOCUMENTATION, & COMPILATION
==============================================================================

This code has been tested using the Java SE 8

This release of EBLA is being distributed as a single JAR file
(EBLA_1.0.jar) containing the source, binaries, and JavaDoc
documentation.

Note that this JAR file only contains the AVI files for the "Animation"
Experience Set. Unless you download the full video dataset (see VIDEO DATASET
below), you will only be able to run EBLA Sessions for the 1st record on the
Parameter Screen.

There is a changelog.txt file, which contains changes for EBLA.

To extract EBLA, place the JAR file where you would like it installed (e.g.
"c:\temp\" or "/home/<username>/") and issue the command:

jar -xf EBLA_1.0.jar

To run EBLA:
  1. navigate to the "scripts" folder/director (e.g. "cd ./ebla/scripts/")
  2. type "winrun.bat" (Windows) or "sh linrun" (Linux)

To recompile EBLA:
  1. navigate to the "scripts" folder/director (e.g. "cd ./ebla/scripts/")
  2. type "wincompile.bat" (Windows) or "sh lincompile" (Linux)
     
To generate the JavaDoc documentation for EBLA:
  1. navigate to the "scripts" folder/director (e.g. "cd ./ebla/scripts/")
  2. type "windocs.bat" (Windows) or "sh lindocs" (Linux)


==============================================================================
DATABASE INSTALLATION
==============================================================================

The EBLA software framework utilizes an embedded H2 database called ebla
for storage of its parameters, dataset, intermediate results, and 
entity-lexeme mappings.


==============================================================================
VIDEO DATASET
==============================================================================

EBLA has been evaluated on a small test set of animations created with 
Macromedia Flash and a much larger test set of real videos.  In both cases,
the files were delivered to EBLA as AVI files.
     
The set of animations are contained in ./experiences/ subdirectory of
the installation directory.  The full set of real videos used to
evaluate EBLA is just over 326MB (compressed JAR file).  It is available 
from the following link:
https://sourceforge.net/projects/ebla/files/EBLA/ebla_experiences.jar/download

To install the full test set, simply download the release file
ebla_experiences.jar to the EBLA installation directory and type:
  jar -xf ebla_experiences.jar
  
The full set of videos will be extracted to the  ./experiences/ subdirectory.    


==============================================================================
EBLA INTERFACE
==============================================================================

Starting with version 0.6.0-alpha, EBLA is run via a graphical user interface
(GUI) developed using Java Swing.  For more information see the user's manual,
"manual.pdf" in the EBLA installation directory.


==============================================================================
RESULTS
==============================================================================

While it is possible to analyze results from EBLA on a run-by-run basis using
the information in the ebla_data database (e.g. the entity_lexeme_data table),
it is generally easier to use the three text files of results generated by
EBLA.

The three text files are delimited using semicolons and are described below:

1. performance.ssv (one record for each pass through the current set of
    experiences):

    loopCount - counter of how many times the entire set of experiences
      has been processed
      
    stdDev - current minimum standard deviation for entity comparisons
    
    runNumber - counter of how many times the current set of experiences
      has been processed for the current minimum standard deviation
      
    experienceIndex - total number of experiences processed during current run
    
    totalSec - seconds elapsed during current run
    
    totalLex - total number of lexemes processed
    
    totalUMLex - total number of lexemes NOT mapped to an entity by EBLA
    
    totalEnt - total number of entities processed
    
    totalUMEnt - total number of entities NOT mapped to a lexeme by EBLA


2. mappings.ssv (one record for each mapped entity in each experience):

    loopCount - counter of how many times the entire set of experiences
      has been processed
      
    experienceIndex - order that experience was processed during current run
    
    resolutionCount - number of experiences elapsed before entity for current
      experience was resolved

      
3. descriptions.ssv (one row for every experience for which a description
    was generated):
    
    loopCount - counter of how many times the entire set of experiences
      has been processed
      
    stdDev - current minimum standard deviation for entity comparisons
    
    experienceIndex - order that experience was processed during current run
    
    generatedDescription - description generated by EBLA for current 
      experience (comma-separated)
      
    numCorrect - number of correctly generated lexemes for current experience
    
    numWrong - number of incorrectly generated lexemes for current experience
    
    numUnknown - number of unknown lexemes for current experience
    
    origDescription - original description for current experience (from
      experience_data table)
      
Note that descriptions.ssv is only generated when EBLA is run in description
mode (e.g. generate_desc_code = 1).


==============================================================================
MISC
==============================================================================

For more information on EBLA, see:
  https://aclanthology.org/W03-0607.pdf
  https://digitalcommons.lsu.edu/gradschool_dissertations/2162/
  
For more information on EDISON, see:
  https://web.archive.org/web/20100814112001/http://coewww.rutgers.edu/riul/research/code/EDISON/index.html

For more information on H2, see:
  https://www.h2database.com
  
For more information on Java, see:
  https://www.java.com
  
For questions or comments regarding EBLA, send e-mail to
Brian E. Pangburn (ebla@greatmindsworking.com)


CHANGELOG.txt
==============================================================================
ChangeLog file for the Experience-Based Language Acquisition (EBLA) system.
==============================================================================

==============================================================================
Version 1.1.0 - Released TBD
==============================================================================

1. 2018-02-09: Migrated EBLA SourceForge Repository from CVS to git.
2. 2021-01-26: Maven conversion/support.
3. 2021-12-23: Updated H2 to 2.0.204 and modified millisecond evaluation in SQL
    (EBLA.java around line 466)
4. 2023-12-15: Moved CHANGELOG entries related to dependencies/plugins into
    CHANGELOG-POMS.txt
5. 2023-12-15: Added explicit casts in PolyAnalyzer around line 130-140 to
    avoid GitHub implicit narrowing alerts.

==============================================================================
Version 1.0 - Released 2014-12-26
==============================================================================

1. Misc. code cleanup

2. Added ./jedison/ folder with script to build separate jEDISON JAR

3. Renamed ./scripts/linbuildjar to ./scripts/linbuildeblajar

4. Made com.greatmindsworking.EDISON.segm.GFunction an abstract class

5. Updates to EBLA & jEDISON readme.txt files

6. Updates to latest H2 jar (h2-1.4.184)

7. readme.txt cleanup

8. improved UI functionality surrounding database connections

==============================================================================
Version 0.9 - Released 06-21-2014
==============================================================================

1. Moved various compilation scripts to /EBLA/scripts

2. Compiling for latest JRE (jdk1.7.0_45)

3. Updates to latest H2 jar (h2-1.4.177)

4. Retooling release jar as a single file with (bin, docs, & src),
   with all extracting to a ./ebla directory


==============================================================================
Version 0.8 - Released 06-10-2011
==============================================================================

1. Added support for Generics throughout

2. Compiling for Java 1.6

3. Addressed all -Xlint warnings (some via @SuppressWarnings)

4. Adding screen default values for new records (TO DO)

5. Switched database from PostgreSQL to H2 embedded database

6. Added Eclipse .project & .classpath files to CVS


==============================================================================
Version 0.7.1-alpha - Released 06-10-2005
==============================================================================

1. Restructured source tree such that all source code is now in /EBLA/src

2. Fixed misc Linux/Windows issues among compilation, execution, etc. scripts

3. Updated GUI to use SwingSet 1.0.0-PR2

4. Updated JDBC driver to postgresql-8.0-311.jdbc3.jar

5. Removed utils.jar and added /src/com/greatmindsworking/utils/DBConnector.java

6. Added linbuildjars script for building various EBLA JAR files


==============================================================================
Version 0.7.0-alpha
==============================================================================

1. Added experimental genetic algorithm utility for training vision system.

2. Pass displayText variable from SessionData through to  segmentation  system
   to control printing of intermediate results.
   
3. Modified segmentation system to operate based on port of original release
   (04-25-2002) of EDISON system or port of revised release (04-14-2003) of
   EDISON system.
   
4. Added edison_port_version variable to ebla_data database SQL (0=04-25-2002, 
   1=04-14-2003) to control #3 above.  Also added to EBLA GUI.
   
5. Modified EBLA.java to flush results to various log file for every run so
   that results can be monitored during execution.


==============================================================================
Version 0.6.0-alpha
==============================================================================

1. Overhaul of ebla_data database to allow storage of results across multiple
   calculation sessions.
   
2. Addition of Swing-based GUI.

3. Addition of User's Manual.


==============================================================================
Version 0.5.0-alpha
==============================================================================

(initial release)

archive/jEDISON/SegTest.java
/* $Id$
 *
 * Tab Spacing = 4
 *
 * Copyright (c) 2002-2003, Brian E. Pangburn & Jonathan P. Ayo
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.  Redistributions in binary
 * form must reproduce the above copyright notice, this list of conditions and
 * the following disclaimer in the documentation and/or other materials
 * provided with the distribution.  The names of its contributors may not be
 * used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * This software is a partial port of the EDISON system developed by
 * Chris M. Christoudias and Bogdan Georgescu at the Robust Image
 * Understanding Laboratory at Rutgers University
 * (http://www.caip.rutgers.edu/riul/).
 *
 * EDISON is available from:
 * http://www.caip.rutgers.edu/riul/research/code/EDISON/index.html
 *
 * It is based on the following references:
 *
 * [1] D. Comanicu, P. Meer: "Mean shift: A robust approach toward feature
 *     space analysis". IEEE Trans. Pattern Anal. Machine Intell., May 2002.
 *
 * [2] P. Meer, B. Georgescu: "Edge detection with embedded confidence".
 *     IEEE Trans. Pattern Anal. Machine Intell., 28, 2001.
 *
 * [3] C. Christoudias, B. Georgescu, P. Meer: "Synergism in low level vision".
 *     16th International Conference of Pattern Recognition, Track 1 - Computer
 *     Vision and Robotics, Quebec City, Canada, August 2001.
 *
 * The above cited papers are available from:
 * http://www.caip.rutgers.edu/riul/research/robust.html
 *
 */



import java.awt.image.*;
import java.io.*;
import javax.imageio.ImageIO;
import com.greatmindsworking.EDISON.segm.*;



/**
 * SegTest.java
 *
 * SegTest is used to demonstrate the mean shift image segmentation code.
 *
 * @author	$Author$
 * @version	$Revision$
 */
public class SegTest {

    /**
     * Main procedure - allows SegTest to be run from the command line
     *
     * The user can pass a graphics file name and the desired segmentation parameters
     * to the main method which will then perform mean shift based image segmentation
     * on that image.
     */
    public static void main(String[] args) {

		// DECLARATIONS
			String inputFile = "./input.png"; 	// INPUT FILE NAME
			String outputFile = "./segment.png";// OUTPUT FILE NAME

			int edisonPortVersion = 1;			// INDICATES WHICH PORT OF EDISON TO USE (0=04-25-2003; 1=04-14-2003)
			boolean displayText = true;			// INDICATES WHETHER OR NOT TO DISPLAY DETAILED MESSAGES
			float colorRadius = (float)6.5; 	// COLOR RADIUS FOR MEAN SHIFT ANALYSIS IMAGE SEGMENTATION
			int spatialRadius = 7; 				// SPATIAL RADIUS FOR MEAN SHIFT ANALYSIS IMAGE SEGMENTATION
			int minRegion = 20; 				// MINIMUM NUMBER OF PIXEL THAT CONSTITUTE A REGION FOR
												// MEAN SHIFT ANALYSIS IMAGE SEGMENTATION
			int speedUp = 1; 					// SPEED-UP LEVEL FOR MEAN SHIFT ANALYSIS IMAGE SEGMENTATION
												// 0=NO SPEEDUP, 1=MEDIUM SPEEDUP, 2=HIGH SPEEDUP
			float highSpeedUpFactor = (float)0.0; // WHEN speedUp = 2, values range between 0.0 (high quality) and 1.0 (high speed)

			boolean warnUser = false;

		try {

			// CHECK TO SEE SIX PARAMETERS WERE PASSED FROM THE COMMAND LINE
				if (args.length >= 7) {
					// EXTRACT SOURCE IMAGE
						inputFile = args[0];

					// EXTRACT VERSION
						edisonPortVersion = Integer.parseInt(args[1]);

					// EXTRACT DISPLAY FLAG - NOTE parseBoolean NOT ADDED UNTIL J2SE 1.5
						displayText = Boolean.getBoolean(args[2]);

					// EXTRACT COLOR RADIUS
						colorRadius = Float.parseFloat(args[3]);

					// EXTRACT SPATIAL RADIUS
						spatialRadius = Integer.parseInt(args[4]);

					// EXTRACT MINIMUM REGION
						minRegion = Integer.parseInt(args[5]);

					// EXTRACT SPEEDUP LEVEL
						speedUp = Integer.parseInt(args[6]);
						if (speedUp == 2) {
							if (args.length==6) {
								highSpeedUpFactor = Float.parseFloat(args[7]);
							} else {
								warnUser=true;
							}
						}


				} else {
					// WARN USER
						warnUser = true;
				}


			// TELL USER HOW TO USE TEST CLASS (IF NECESSARY)
				if (warnUser) {
				// DISPLAY MESSAGE
					System.out.println("Usage: java SegTest <source image> <version: 0=04-25-2002, 1=04-14-2003>"
					    + " <display: true/false> <color radius> <spatial radius>"
						+ " <min region> <speedup: 0=none, 1=medium, 2=high> <for speedup of 2, speed factor: 0.0=better quality, 1.0=better speed>");
					System.out.println("");
					System.out.println("e.g.  java SegTest my_image.png 0 true 6.5 7 20 0");
					System.out.println(" -or- java SegTest my_image.png 1 false 6.5 7 20 1");
					System.out.println(" -or- java SegTest my_image.png 1 true 6.5 7 20 2 0.5");
					System.out.println("");
					System.out.println("For more information on mean shift image segmentaton, consult:");
					System.out.println(" [1] D. Comanicu, P. Meer: 'Mean shift: A robust approach toward feature");
					System.out.println("     space analysis'. IEEE Trans. Pattern Anal. Machine Intell., May 2002.");
					System.out.println("");
					System.out.println(" [2] P. Meer, B. Georgescu: 'Edge detection with embedded confidence'.");
					System.out.println("     IEEE Trans. Pattern Anal. Machine Intell., 28, 2001.");
					System.out.println("");
					System.out.println(" [3] C. Christoudias, B. Georgescu, P. Meer: 'Synergism in low level vision'.");
					System.out.println("     16th International Conference of Pattern Recognition, Track 1 - Computer");
					System.out.println("     Vision and Robotics, Quebec City, Canada, August 2001.");
					System.out.println("");
					System.out.println(" The above cited papers are available from:");
					System.out.println(" http://www.caip.rutgers.edu/riul/research/robust.html");

				// EXIT
					System.exit(0);
				}


			// LOAD SOURCE IMAGE USING LOADER CLASS
				BufferedImage tmpImage = ImageIO.read(new File(inputFile));

			// DETERMINE WIDTH AND HEIGHT
				int width = tmpImage.getWidth();
				int height = tmpImage.getHeight();

			// CROP IMAGE
			//	tmpImage = tmpImage.getSubimage(5, 5, width-5, height-5);

			// RECALCULATE WIDTH AND HEIGHT
				width = tmpImage.getWidth();
				height = tmpImage.getHeight();

			// DETERMINE NUMBER OF PIXELS
				int pixelCount = width * height;

			// INITIALIZE ARRAYS FOR RGB PIXEL VALUES
				int rgbPixels[] = new int[pixelCount];
				tmpImage.getRGB(0, 0, width, height, rgbPixels, 0, width);

			// CREATE MSImageProcessor OBJECT
				MSImageProcessor mySegm = new MSImageProcessor();

			// SET IMAGE
				mySegm.DefineBgImage(rgbPixels, ImageType.COLOR, height, width);

			// SET SetSpeedThreshold FOR HIGH SPEEDUP OPTION
				if (speedUp == 2) {
					mySegm.SetSpeedThreshold(highSpeedUpFactor);
				}

			// SEGMENT IMAGE
				if (speedUp == 0) {
					mySegm.Segment(edisonPortVersion, displayText, spatialRadius,
						colorRadius, minRegion, SpeedUpLevel.NO_SPEEDUP);
				} else if (speedUp == 1) {
					mySegm.Segment(edisonPortVersion, displayText, spatialRadius,
						colorRadius, minRegion, SpeedUpLevel.MED_SPEEDUP);
				} else {
					mySegm.Segment(edisonPortVersion, displayText, spatialRadius,
						colorRadius, minRegion, SpeedUpLevel.HIGH_SPEEDUP);
				}

			// GET RESULTING SEGMENTED IMAGE (RGB) PIXELS
				int segpixels[] = new int[pixelCount];
				mySegm.GetResults(segpixels);

			// BUILD BUFFERED IMAGE FROM RGB PIXEL DATA
				BufferedImage segImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
				segImage.setRGB(0, 0, width, height, segpixels, 0, width);

			// SAVE BUFFERED IMAGE(S) AS PNG
				ImageIO.write(segImage, "png", new File(outputFile));

			// EXIT
				System.exit(0);

		} catch (Exception e) {
			System.out.println("\n--- SegTest.main() Exception ---\n");
			e.printStackTrace();
		}

    } // end main()

}; // end SegTest class



/*
 * $Log$
 * Revision 1.7  2004/06/24 15:19:15  yoda2
 * Replaced parseBoolean with getBoolean because parseBoolean only works with J2SE 1.5.
 *
 * Revision 1.6  2004/03/02 03:27:44  yoda2
 * Updated for 0.7.0 release of jEDISON & EBLA.  Added port version option & detailed display option.
 *
 * Revision 1.5  2003/11/24 16:46:04  yoda2
 * Updated usage instructions.
 *
 * Revision 1.4  2003/11/24 16:19:29  yoda2
 * Updated copyright to 2002-2003.
 *
 * Revision 1.3  2003/11/24 14:47:48  yoda2
 * Fixed bug that was setting the color and spatial radii to the value supplied for the spatial radius.  Added a factor to control speed (1.0) vs. quality (0.0) for the high speedup option (added for 4-14-2003 release of C++ EDISON).
 *
 * Revision 1.2  2002/12/11 22:37:52  yoda2
 * Initial migration to SourceForge.
 *
 * Revision 1.1  2002/09/20 19:48:04  bpangburn
 * Added stand-alone test class for EDISON port to CVS.
 *
 *
 */